### ISSUE TITLE
Remove all unnecessary custom VJPs for differentiable control flow.

### ISSUE THREAD

**Issue by rxwei** [tensorflow/swift-apis#389]
https://github.com/apple/swift/pull/26355 fixed all control flow memory leaks (#364). After that patch lands, we should remove all unnecessary custom VJPs for differentiable control flow, such as the custom VJP for `BatchNorm.callAsFunction(_:)`.

**Comment by eaplatanios** [comment]
I'll take this on because I also want to check if the leaks in my RL library have been fixed.

**Comment by eaplatanios** [comment]
@rxwei even though the situation improved after apple/swift#26355, there are still memory leaks somewhere related to auto-diff of control flow. For example, using instruments I see that there are leaks related to `Differentiable.move` coming from `Dense.AllDifferentiableVariables.+`. I also see that `Conv2D.AllDifferentiableVariables.+` results in much less memory leaked, suggesting that the `if` statement in the dense layer may be the culprit.

**Comment by rxwei** [comment]
It'd be helpful if you can share a stack trace. It's hard to tell whether this is a problem with AD control flow especially when it has something to do with `move(along:)`. Also, could you please provide your compiler version by obtaining it from `swift --version`?

**Comment by eaplatanios** [comment]
Swift version is:

```bash
Swift version 5.1-dev (LLVM 200186e28b, Swift be21761f8f)
Target: x86_64-apple-darwin18.6.0
```

and screenshot of a stacktrace:

<img width="1321" alt="Screen Shot 2019-07-14 at 7 25 12 PM" src="https://user-images.githubusercontent.com/1294940/61925621-1dd5e880-af3b-11e9-8cfc-c647c16674ab.png">

Does this help?

**Comment by eaplatanios** [comment]
@rxwei sorry I uploaded the wrong image earlier. I just edited the post with the correct image.

**Comment by rxwei** [comment]
Ok. This shouldn't have anything to do with differentiation. `move(along:)` albeit part of `Differentiable` is just an API and is never used by automatic differentiation. Plus, this stack trace show no AD-generated code.

**Comment by eaplatanios** [comment]
So, what do you think is causing this? This only happens during the model update so it seems to have to do with differentiation somehow.

**Comment by rxwei** [comment]
Optimizer APIs are not using differentiation, so it's more likely the case that you have a reference cycle somewhere. I'm be happy to help you debug, but I haven't been able to build retro-swift. I'll try again.

**Comment by eaplatanios** [comment]
Thanks Richard! You don't have to build retro-swift actually. I recently created an API for the arcade learning environment that is much faster. I don't have too detailed instructions yet, but if you clone [this repo](https://github.com/eaplatanios/swift-ale) and follow the instructions in the readme file, then hopefully you'll be able to just do `swift run` afterwards. Please let me know if that works for you and in the meantime I'll also try using SGD or some other simpler optimizer and see if the leak persists. :)

**Comment by rxwei** [comment]
It seems to rely on #346.

**Comment by eaplatanios** [comment]
Oops. Yes it does. If you add a dependency to the branch corresponding to that PR in SwiftPM it may work. I haven't tried overriding the stdlib-provided `TensorFlow`.

**Comment by eaplatanios** [comment]
The leak exists with SGD too btw.

**Comment by rxwei** [comment]
Let's move the discussion to a separate issue in your library repo? It's highly unlikely that this is related to automatic differentiation and requires further investigation. I'm happy to build a toolchain based on #346 and help debug.

**Comment by eaplatanios** [comment]
Thanks Richard! Yes, moving it there sounds good. I’m going to bed now but let’s continue the discussion tomorrow. FYI, the example in the repo does train if you let it run, but the leak becomes too bad after about 500-1000 steps making it unusable on my MacBook.

**Comment by rxwei** [comment]
Update: Confirmed in eaplatanios/swift-ale#1 that the memory leak is unrelated to AD.
